Added three_leg_turnaround generator

mjrand opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
I've added in a new python file to generate three leg turnarounds and modified the `generate_constant_velocity_scan` in the ACU `drivers.py` to accept the new turnaround function.

## Description
<!--- Describe your changes in detail -->
This function is the implementation of the three part turnarounds discussed on last weeks SAT Analysis call.
This turnaround function breaks the turnarounds into three "legs":

1. The initial deceleration.
2. The middle portion with a gentle/zero velocity acceleration to allow the motor gears to change direction and come in contact with the opposite azimuth bearing face more gently.
3. The final acceleration to the scan velocity.

This should hopefully make the turnarounds more gentle for the SATs. 

I have renamed the turnarounds from "full_stop_turnarounds" to "three_leg_turnarounds". I am not married to this naming scheme and I am OK with changing any names in this PR prior to approval. 

A few other points of discussion:
1. This current implementation forces the time of the second leg of the turnaround to be 1 second. This limits the minimum turnaround time to ~2 seconds (so that there is at least 0.5 seconds for the first and third leg). This hard coded 1 second is only because that is the only time I've tested for the second leg. This is really only a problem for planet scans where the turnaround time is only (2 * 0.8 deg/s) / 1.0deg/s^2 = 1.6 seconds. I think the solution to this is testing an implementation where each leg has an equal length, but this would need to be tested.
2. The step_time for the turnaroudn is hard coded to 0.1seconds. It may not need to be this fast but I don't think there's a harm in keeping it this fast. Because I have to control the ACU very directly through the turnaround there is probably a maximum step time to keep the ACU following the three leg turnaround accurately but I haven't tested it.
3. I've implemented this with a "TrackPoint queue" to ensure the batch_size is always obeyed. Because the three_leg_turnaround function returns a list of TrackPoints at the beginning of the turnaround I had to either fully ignore the batch_size when turnarounds happened or implement this queue. Matthew has said we don't have to obey the batch_size so we can ditch the queue for a slightly simpler implementation if desired.

This implementation basically leaves 95% of the `constant_scan_velocity` generator the same and only changes the turnaround portions. I've added a simple boolean into the function call that activates the new turnaround. If it is set to False then the constant velocity scan should generate the same trajectories as before.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This implements the new three leg turnarounds for the SATs that have shown improvements to the FPA temps during turnarounds.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

These turnarounds have been tested on SATp1 multiple times. We should probably test that this specific implementation of the `constant_velocity_scan` generator function works!

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

mhasself commented on socs/agents/acu/three_leg_turnaround.py:
```diff
@@ -0,0 +1,310 @@
+from socs.agents.acu.drivers import TrackPoint
+
+
+def gen_three_leg_turnaround(t0, az0, el0, v0, turntime, az_flag, el_flag, point_group_batch, **kwargs):
```
Why are kwargs not declared explicitly here?

mjrand replied:
I've changed them to be explicit since there's only three of them.

mhasself commented on socs/agents/acu/three_leg_turnaround.py:
```diff
@@ -0,0 +1,310 @@
+    ts = np.arange(t_i, t_f, step_time)  # Divide our times into points with step_time spacing
+    vs = x0(ts, *A)  # Solve for the velocity at every time using our solved components
+
+    an1 = get_an1(t_i, xn1_i, *A)  # We have to solve for the first antiderivative component to get the positions
+    xs = xn1(ts, an1, *A)  # Now solve for the positions at every time
```
I think these two can be replaced by
```
xs = xn1(ts, 0, *A)
xs = xs - xs[0] + xn1_i
```
Then you don't need the `get_an1` function.

mjrand replied:
Oh yeah that does work... and its so much more elegant! Awesome, thanks!

mhasself commented on socs/agents/acu/three_leg_turnaround.py:
```diff
@@ -0,0 +1,310 @@
+    return an1 + a0 * t + (1 / 2) * a1 * t**2 + (1 / 3) * a2 * t**3 + (1 / 4) * a3 * t**4 + (1 / 5) * a4 * t**5 + (1 / 6) * a5 * t**6
+
+
+def x0(t, a0, a1, a2, a3, a4, a5):
```
I believe the functions below this point aren't used?  Remove.

mhasself reviewed (changes requested):
Couple specific changes requested.

We also need this functionality controllable from the generate_scan Process --  we can't just turn this on, by default, for everyone, forever.

I'd have thought this needed to be run in "free form" mode... was that not the case?

mjrand commented on socs/agents/acu/drivers.py:
```diff
@@ -278,7 +280,8 @@ def generate_constant_velocity_scan(az_endpoint1, az_endpoint2, az_speed,
                                     az_first_pos=None,
-                                    az_drift=None):
+                                    az_drift=None,
+                                    three_leg_turnaround=True):
```
Makes sense to me. I've changed the argument to `turnaround_method=None` and added in the following if statement where the turnarounds are generated:
```if turnaround_method is not None and turnaround_method == "three_leg":```

This will preserve the default behavior unless specified by `turnaround_method`.

mjrand commented on socs/agents/acu/three_leg_turnaround.py:
```diff
@@ -0,0 +1,310 @@
+        el_flag (int): The el flag used by the ACU. Inherited from the scan generation function and not changed. Used for TrackPoints.
+        point_group_batch (int): the point group batch used by the ACU. Inherited from the scan generation function and not changed. Used for TrackPoints.
+
+    kwargs:
```
Is `**kwargs` usually not the correct way to use them? I've just assumed them to be these sets of optional arguments but maybe there's a better way to do this stylistically. (I could just add these three args into the function header with default values if that seems better)

mjrand commented on socs/agents/acu/three_leg_turnaround.py:
```diff
@@ -0,0 +1,310 @@
+        point_group_batch (int): the point group batch used by the ACU. Inherited from the scan generation function and not changed. Used for TrackPoints.
+
+    kwargs:
+        midpoint_time (float): The time used by the second leg of the turnaround. Defaults to 1 second.
```
The second leg of the turnaround here isn't necessarily "stationary" of the velocity is non-zero... how about "second_leg_time"/"second_leg_velocity"?

mjrand commented:
> Couple specific changes requested.

Addressed all of the comments!

> We also need this functionality controllable from the generate_scan Process -- we can't just turn this on, by default, for everyone, forever.

I've added in changes to the agent as well!

> I'd have thought this needed to be run in "free form" mode... was that not the case?

I've never dealt with "free form" mode... I just ran it using `from_file` and pre-generated scripts. Should I be using free form mode? I actually don't know to be honest.

mjrand commented:
Closing this and opening a new PR with a new branch here : #931

The commit is identical to this commit but the merge that added in type2 and type3 scans into the acu driver diverged my branch in a way that I didn't want to learn how to fix... so I just made a new branch based off of main and recommitted.
